### PR TITLE
added fetchX smarty functions for db queries from template

### PR DESCRIPTION
**Pull request by dneustadt**
Fetch data from the database using the smarty functions:

* `fetchOne` - fetches a single value from the specified column of a row
* `fetchRow` - fetches the first row from the specified table
* `fetchAll` - fetches all rows from the specified table

The following parameters are available

* `var` - required for `fetchRow` and `fetchAll`, defines the variable the resulting array will be stored in
* `select` - may either be of type `string` or `array`, specifies the columns to be fetch, required for `fetchOne`
* `from` - required for all functions, specifies the table to be fetched from
* `where` - must be of type `array`, must contain key value pairs in the format of `'[column][operand]' => value` 
* `order` - must be of type `array`, must contain key value pairs in the format of `'[column]' => '[ASC][DESC]'` 
* `offset` - must be of type `integer`, sets the offset for the first result
* `limit` - must be of type `integer`, sets the limit of results

Examples

```smarty
{* Fetches the price of the last article added to basket of the current user *}
{fetchOne select="price" from="s_order_basket" where=["modus =" => 0, "sessionID =" => $smarty.session.Shopware.sessionId] order=["id" => "DESC"]}

{* Fetches the name and description of the 5th to 15th article *}
{fetchAll var="articles" from="s_articles" offset=5 limit=10}
{foreach $articles as $article}
    Article: {$article.name} - {$article.description}<br>
{/foreach}

{* Fetches the last order's id and ordernumber *}
{fetchRow var="order" select=["ordernumber", "id"] from="s_order" where=["status !=" => -1] order=["ordertime" => "DESC"]}
{$order.id} - {$order.ordernumber}
```

### CODE REVIEW

**Comment by JoshuaBehrens** [comment]
Although I dislike having queries in the template it will help out a lot for small solutions.

I do not like the way how the where expressions are made (or I misunderstand them). If I read it correctly the where array is understood as key value pair whereas key is the prefix and value is a bound parameter. How should an in-expression be written? How should I compare column1 with column2?

I'd like to see the exceptions be wrapped in a FroshTemplateExtension specific exception to have it more obvious in an exception occurence that is caused by this plugin.

**Comment by dneustadt** [comment]
You can use in-expressions like this:
```
{fetchAll var="articles" from="s_articles" where=["id IN" => [3, 6, 9]]}
```
I realize it's a little clunky to have the column and expression within a single string. I just fear that abstracting expressions further might lead to an overblown documentation when each expression needs to be documented. Also it should be noted that this whole feature needs to be treated with care, obviously don't use user inputs for setting column and expression etc.

**Comment by dneustadt** [comment]
As for using two columns in an expression, I would suggest omitting the value and writing the whole expression in the key e.g. `['column1 = column2' => null]`

File: Extensions/Fetch/FetchService.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var array
+     */
+    private $_cache = [];
```
Comment (shyim) [inline Extensions/Fetch/FetchService.php:22]: Why the property begins with underscore? :sweat_smile:

**Comment by shyim** [comment]
I have fear what template developers does with that. I am thinking about a case, where the theme developer could open a new security issue

File: Extensions/Fetch/FetchService.php (hunk 1 above)
Comment (dneustadt) [reply Extensions/Fetch/FetchService.php:22]: No actual reason 😂 Derives from an outdated naming convention for private properties. I still like to use it for private properties that provide no functional use and contain pure data.

**Comment by dneustadt** [comment]
Yes, I knew you would say that because I know your stance on that issue. 😂  But may I ask you to provide an example? The only real risk I see would stem from theme developers compiling raw user inputs as smarty syntax. I mean, you'd have to be pretty stupid to do this anyway. While giving away the contents of your database at this point (which, of course, would be worse enough), at least you wouldn't be able to inject malicious sql. I tried to break it in every way but couldn't get any statements to be executed aside from the select statement. And keep in mind, to even make this possible you'd have to be pretty stupid and allow raw user inputs to be compiled in smarty. So I see no risk other than utter stupidity and at that point nobody could help you anyway. 😂

**Comment by JoshuaBehrens** [comment]
Those who are stupid do not know this. So you can't just say: the template engineers should not be stupid when using this. Imagine a frontend designer just fixing a thing in a small agency. Sees this plugin in the shop and sees that it can be used to read somethings from the database. I can see it happening.

An other topic. What about caching? Can I break caches so data will always be loaded from the database? Or the other way around? How can I ensure it is cached?

**Comment by dneustadt** [comment]
Well, I disagree, I can't provide software for toddlers. I'm talking about people that think `{include file="string:"|cat:$smarty.request.foobar}` is a good idea. So the vast majority of users that might benefit from a feature should do without it because of a hypothetical idiot? You can only go so far protecting users from themselves before becoming paranoid and removing any kind of customization.

About caching. Queries that are identical, judging by the provided parameters, are being "cached" during runtime only. For example: Lets say you iterate over products and want to fetch the name of the related category by id. Some products will share the same category id so I execute that exact query only once, store the result and return it on the next iteration that includes the same category id within the service.
So there is no cache between requests and I think that should be expected, since I think it's the point of querying the database directly to retrieve live data. You wouldn't cache queries you fire within your controller actions, right? If you want and need caching, that's what the http cache is for.

**Comment by JoshuaBehrens** [comment]
That is fair point. I guess then there should be at least a hint in the readme that SQL injection is a thing and it is possible with this plugin to have ones if missused. Eventually you can still say: RTFM ;)

Sorry for using the term cache in an abstract way. I was thinking about the http cache. I expect it to use the http cache but can I also break the http cache with just your extension? Or do I have to make my template an action then?

**Comment by dneustadt** [comment]
@shyim how about an option in the plugin configuration to enable/disable the fetchX functions coupled with a tooltip explaining potential risks? could be disabled by default, so theme developers that feel insecure about these functions could leave them disabled.

**Comment by shyim** [comment]
@dneustadt Would be nice :blush:

**Pull request merged**